bqskit/compiler/compiler.py

```diff
@@ -99,108 +117,341 @@ def __exit__(self, type: Any, value: Any, traceback: Any) -> None:
+                self.conn.close()
+            except Exception as e:
+                _logger.debug(
+                    'Unsuccessfully disconnected from runtime server.',
+                )
+                _logger.debug(e)
+            else:
+                _logger.debug('Disconnected from runtime server.')
+            finally:
+                self.conn = None
+
+        # Shutdown server if attached
+        if self.p is not None and self.p.pid is not None:
+            try:
+                os.kill(self.p.pid, signal.SIGINT)
+                _logger.debug('Interrupted attached runtime server.')
+
+                self.p.join(1)
+                if self.p.exitcode is None:
+                    os.kill(self.p.pid, signal.SIGKILL)
+                    _logger.debug('Killed attached runtime server.')
+
+            except Exception as e:
+                _logger.debug(
+                    f'Error while shuting down attached runtime server: {e}.',
+                )
+            else:
+                _logger.debug('Successfully shutdown attached runtime server.')
+            finally:
+                self.p.join()
+                _logger.debug('Attached runtime server is down.')
+                self.p = None
+
+        # Reset interrupt signal handler and remove exit handler
+        signal.signal(signal.SIGINT, self.old_signal)
+        atexit.unregister(self.close)
 
     def __del__(self) -> None:
-        if self.managed and hasattr(self, 'processes'):
-            for proc in self.processes:
-                proc.terminate()  # type: ignore
-                proc.wait()  # type: ignore
-
-    def submit(self, task: CompilationTask) -> None:
-        """Submit a CompilationTask to the Compiler."""
-        executor = self.client.scatter(Executor(task))
-        future = self.client.submit(Executor.run, executor, pure=False)
-        self.tasks[task.task_id] = future
-        _logger.info('Submitted task: %s' % task.task_id)
-
-    def status(self, task: CompilationTask) -> str:
-        """Retrieve the status of the specified CompilationTask."""
-        return self.tasks[task.task_id].status
-
-    def result(self, task: CompilationTask) -> Circuit:
-        """Block until the CompilationTask is finished, return its result."""
-        circ = self.tasks[task.task_id].result()[0]
-        return circ
-
-    def cancel(self, task: CompilationTask) -> None:
-        """Remove a task from the compiler's workqueue."""
-        self.client.cancel(self.tasks[task.task_id])
-        _logger.info('Cancelled task: %s' % task.task_id)
-
-    def compile(self, task: CompilationTask) -> Circuit:
-        """Submit and execute the CompilationTask, block until its done."""
-        _logger.info('Compiling task: %s' % task.task_id)
-        self.submit(task)
-        result = self.result(task)
-        return result
+        self.close()
+        _logger.debug('Compiler successfully shutdown.')
+
+    def submit(
```

Comment:
Why not split the API to submmit_task and submmit_workflow?

Reply:
For backward compatibility.